### ISSUE TITLE
tf.GradientTape training much slower than keras.fit

### ISSUE THREAD

**Issue by bbreton3** [tensorflow/tensorflow#33898]
<em>Please make sure that this is a bug. As per our [GitHub Policy](https://github.com/tensorflow/tensorflow/blob/master/ISSUES.md), we only address code/doc bugs, performance issues, feature requests and build/installation issues on GitHub. tag:bug_template</em>

**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow):
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Linux Ubuntu 1*.04
- TensorFlow version (use command below): 2.0.0
- Python version: 3.7.4
- CUDA/cuDNN version: 10.0
- GPU model and memory: GTX 1080

I am currently trying to get a hold of the **TF2.0** api, but as I compared the [**GradientTape**][1] to a regular **keras.Model.fit** I noticed: 

1. It ran slower(probably due to the Eager Execution)

2. It converged much slower (and I am not sure why).

```
+--------+--------------+--------------+------------------+
|  Epoch | GradientTape | GradientTape | keras.Model.fit  |
|        |              |  shuffling   |                  |
+--------+--------------+--------------+------------------+
|    1   |     0.905    |     0.918    |      0.8793      |
+--------+--------------+--------------+------------------+
|    2   |     0.352    |     0.634    |      0.2226      |
+--------+--------------+--------------+------------------+
|    3   |     0.285    |     0.518    |      0.1192      |
+--------+--------------+--------------+------------------+
|    4   |     0.282    |     0.458    |      0.1029      |
+--------+--------------+--------------+------------------+
|    5   |     0.275    |     0.421    |      0.0940      |
+--------+--------------+--------------+------------------+
```

Here is the training loop I used with the **GradientTape**:
```python

optimizer = keras.optimizers.Adam()
glove_model = GloveModel(vocab_size=len(labels))
train_loss = keras.metrics.Mean(name='train_loss')

@tf.function
def train_step(examples, labels):
	with tf.GradientTape() as tape:
		predictions = glove_model(examples)
		loss = glove_model.glove_loss(labels, predictions)

	gradients = tape.gradient(loss, glove_model.trainable_variables)
	optimizer.apply_gradients(zip(gradients, glove_model.trainable_variables))

	train_loss(loss)



total_step = 0
for epoch in range(epochs_number):

	pbar = tqdm(train_ds.enumerate(), total=int(len(index_data) / batch_size) + 1)

	for ix, (examples, labels) in pbar:

		train_step(examples, labels)


	print(f"Epoch {epoch + 1}, Loss {train_loss.result()}")

	# Reset the metrics for the next epoch
	train_loss.reset_states()
```

And here is the **Keras.Model.fit** training:
```python
glove_model.compile(optimizer, glove_model.glove_loss)
glove_model.fit(train_ds, epochs=epochs_number)
```
Here is the **tf.data.Dataset** source 

```python
train_ds = data.Dataset.from_tensor_slices(
	(np.hstack([index_rows.reshape(-1, 1), index_cols.reshape(-1, 1)]), index_data)
).shuffle(100000).batch(batch_size, drop_remainder=True)
```

And Here is the model.

```python
class GloveModel(keras.Model):

    def __init__(self, vocab_size, dim=100, a=3/4, x_max=100):
        super(GloveModel, self).__init__()

        self.vocab_size = vocab_size
        self.dim = dim
        self.a = a
        self.x_max = x_max

        self.target_embedding = layers.Embedding(
            input_dim=self.vocab_size, output_dim=self.dim, input_length=1, name="target_embedding"
        )
        self.target_bias = layers.Embedding(
            input_dim=self.vocab_size, output_dim=1, input_length=1, name="target_bias"
        )

        self.context_embedding = layers.Embedding(
            input_dim=self.vocab_size, output_dim=self.dim, input_length=1, name="context_embedding"
        )
        self.context_bias = layers.Embedding(
            input_dim=self.vocab_size, output_dim=1, input_length=1, name="context_bias"
        )

        self.dot_product = layers.Dot(axes=-1, name="dot")

        self.prediction = layers.Add(name="add")
        self.step = 0

    def call(self, inputs):

        target_ix = inputs[:, 0]
        context_ix = inputs[:, 1]

        target_embedding = self.target_embedding(target_ix)
        target_bias = self.target_bias(target_ix)

        context_embedding = self.context_embedding(context_ix)
        context_bias = self.context_bias(context_ix)

        dot_product = self.dot_product([target_embedding, context_embedding])
        prediction = self.prediction([dot_product, target_bias, context_bias])

        return prediction

    def glove_loss(self, y_true, y_pred):

        weight = tf.math.minimum(
            tf.math.pow(y_true/self.x_max, self.a), 1.0
        )
        loss_value = tf.math.reduce_mean(weight * tf.math.pow(y_pred - tf.math.log(y_true), 2.0))

        return loss_value



```
I tried multiple configurations and optimizers but nothing seems to change the convergence rate.
I tried to **reshuffle** the dataset but with no luck. I also consulted the different issues on GitHub regarding this problem but they all seem to relate to a `tf.function` miss-use.
I also posted this issue on [**StackOverflow**][2] but I had no luck either.

  [1]: https://www.tensorflow.org/tutorials/quickstart/advanced
  [2]: https://stackoverflow.com/questions/58584359/gradientape-convergence-much-slower-than-keras-model-fit

**Comment by bbreton3** [comment]
The problem came from the **shuffling** using the **tf.Dataset** method. It only shuffled through the dataset one bucket at the time. Using the **Keras.Model.fit** yielded better results because it probably adds another shuffling. 

I added a shuffling with `numpy.random.shuffle` and it improved the performance with both training methods:

The generation of the dataset is now:
```python
numpy_data = np.hstack([index_rows.reshape(-1, 1), index_cols.reshape(-1, 1), index_data.reshape(-1, 1)])

np.random.shuffle(numpy_data)

indexes = np.array(numpy_data[:, :2], dtype=np.uint32)
labels = np.array(numpy_data[:, 2].reshape(-1, 1), dtype=np.float32)

train_ds = data.Dataset.from_tensor_slices(
	(indexes, labels)
).shuffle(100000).batch(batch_size, drop_remainder=True)
``` 

And the results are:

```
+--------+--------------+------------------+
|  Epoch | GradientTape |  keras.Model.fit |
+--------+--------------+------------------+
|    1   |     0.294    |      0.294       |
+--------+--------------+------------------+
|    2   |     0.111    |      0.110       |
+--------+--------------+------------------+
|    3   |     0.089    |      0.089       |
+--------+--------------+------------------+
|    4   |     0.074    |      0.075       |
+--------+--------------+------------------+
|    5   |     0.063    |      0.063       |
+--------+--------------+------------------+
```

The training type per epoch is roughly the same at **2minutes per epoch**.

**Comment by ucasiggcas** [comment]
how many/much the buffer_size is good for training ?
thx
